### ISSUE TITLE
ctx_size = ctx_size + 2, What is the purpose of +2

### ISSUE THREAD

**Issue by MasterKID223** [seeyourmind/TKGElib#6]
I unstand that, to get ECE single list with query(s, p, ?, t) tokens, `ctx_size + 1` is enough, why apply  `ctx_size + 2`。
And in this code:  `attention_mask[:,2:] = attention_mask[:, 2:] & ctx_random_mask & gt_mask` , you start doing token masking from `[:, 2:]`，but in the recover code, you recover masked token from `[:, 1:]`。
The recover code:
`src = src[attention_mask[:, 1:]]` and `src[attention_mask[:, 1:]] = out`。

https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L112
https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L147
https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L163

**Comment by Kizzzy** [comment]
> I unstand that, to get ECE single list with query(s, p, ?, t) tokens, `ctx_size + 1` is enough, why apply `ctx_size + 2`。 And in this code: `attention_mask[:,2:] = attention_mask[:, 2:] & ctx_random_mask & gt_mask` , you start doing token masking from `[:, 2:]`，but in the recover code, you recover masked token from `[:, 1:]`。 The recover code: `src = src[attention_mask[:, 1:]]` and `src[attention_mask[:, 1:]] = out`。
> 
> https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L112 https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L147 https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L163

cls + query + context(50) = 50+2

**Comment by Kizzzy** [comment]
> I unstand that, to get ECE single list with query(s, p, ?, t) tokens, `ctx_size + 1` is enough, why apply `ctx_size + 2`。 And in this code: `attention_mask[:,2:] = attention_mask[:, 2:] & ctx_random_mask & gt_mask` , you start doing token masking from `[:, 2:]`，but in the recover code, you recover masked token from `[:, 1:]`。 The recover code: `src = src[attention_mask[:, 1:]]` and `src[attention_mask[:, 1:]] = out`。
> 
> https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L112 https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L147 https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L163

becasuse eceformer is derived from hitter
ctx_size = ctx_size + 2, because:

in hitter:
second stage(context mix): global_cls + query + context(50) = 52

but in eceformer:
in the first stage(ECER), there query + context(50) = 51, so
src = src[attention_mask[:, 1:]] and src[attention_mask[:, 1:]] = out

but at the second stage(context mix) in eceformer:
author use a new module to replace the transformer module in hitter,
so the first mask in attention_mask is not use in the eceformer

**Comment by MasterKID223** [comment]
> > I unstand that, to get ECE single list with query(s, p, ?, t) tokens, `ctx_size + 1` is enough, why apply `ctx_size + 2`。 And in this code: `attention_mask[:,2:] = attention_mask[:, 2:] & ctx_random_mask & gt_mask` , you start doing token masking from `[:, 2:]`，but in the recover code, you recover masked token from `[:, 1:]`。 The recover code: `src = src[attention_mask[:, 1:]]` and `src[attention_mask[:, 1:]] = out`。
> > https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L112 https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L147 https://github.com/seeyourmind/TKGElib/blob/main/kge/model/eceformer.py#L163
> 
> becasuse eceformer is derived from hitter ctx_size = ctx_size + 2, because:
> 
> in hitter: second stage(context mix): global_cls + query + context(50) = 52
> 
> but in eceformer: in the first stage(ECER), there query + context(50) = 51, so src = src[attention_mask[:, 1:]] and src[attention_mask[:, 1:]] = out
> 
> but at the second stage(context mix) in eceformer: author use a new module to replace the transformer module in hitter, so the first mask in attention_mask is not use in the eceformer

Thanks for your explanation, it works.
